File: daily/Linked List/146-LRU-cache.py
class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}

        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def remove(self, node: Node):
        prev_node = node.prev
        next_node = node.next

        prev_node.next = next_node
        next_node.prev = prev_node

    def add(self, node: Node):
        head_first = self.head.next
        node.next = head_first
        node.prev = self.head
        self.head.next = node
        head_first.prev = node


    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.remove(node)
            self.add(node)
            return node.value
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.remove(self.cache[key])
        new_node = Node(key, value)
        self.cache[key] = new_node
        self.add(new_node)

        if len(self.cache) > self.capacity:
            lru_node = self.tail.prev
            self.remove(lru_node)
            del self.cache[lru_node.key]

File: daily/Linked List/146-LRU-cache.go
package daily

type Node struct {
	Key   int
	Value int
	Prev  *Node
	Next  *Node
}

type LRUCache struct {
	Capacity int
	Cache    map[int]*Node
	Head     *Node
	Tail     *Node
}

func Constructor(capacity int) LRUCache {
	head := &Node{Key: 0, Value: 0}
	tail := &Node{Key: 0, Value: 0}
	head.Next = tail
	tail.Prev = head
	return LRUCache{
		Capacity: capacity,
		Cache:    make(map[int]*Node),
		Head:     head,
		Tail:     tail,
	}
}

func (this *LRUCache) remove(node *Node) {
	prevNode := node.Prev
	nextNode := node.Next

	prevNode.Next = nextNode
	nextNode.Prev = prevNode
}

func (this *LRUCache) add(node *Node) {
	headFirst := this.Head.Next
	node.Next = headFirst
	node.Prev = this.Head
	this.Head.Next = node
	headFirst.Prev = node
}

func (this *LRUCache) Get(key int) int {
	if node, ok := this.Cache[key]; ok {
		this.remove(node)
		this.add(node)
		return node.Value
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if node, ok := this.Cache[key]; ok {
		this.remove(node)
	}
	newNode := &Node{Key: key, Value: value}
	this.Cache[key] = newNode
	this.add(newNode)

	if len(this.Cache) > this.Capacity {
		lruNode := this.Tail.Prev
		this.remove(lruNode)
		delete(this.Cache, lruNode.Key)
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
